### PR TITLE
Remove special chars from xml output

### PR DESCRIPTION
**Pull request by slawekjaranowski**

### CODE REVIEW

**Comment by slawekjaranowski** [comment]
related to: https://github.com/apache/maven-invoker-plugin/pull/242

File: src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java [hunk 1]
```diff
@@ -82,6 +82,8 @@ private String createExpectedXML(boolean escape) {
         buf.append(LS);
         buf.append("  </el6>");
         buf.append(LS);
+        buf.append("  <el8>special-char-" + (char) 7 + "</el8>");
```
Comment (slawekjaranowski) [inline src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java:85]: @michael-o @elharo
special char here should be removed or replaced .... escaping is not enough
it is forrbiden char in xml ....

Comment (michael-o) [reply src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java:85]: I'd like to have a reference for that. I believe that you are right, but I also remember that entities are permissive.

**Comment by michael-o** [comment]
Looking at the code as a whole I absolutely do not understand why the framework allows to produce invalid (unescaped) XML. It simply does not make sense to me.

**Comment by michael-o** [comment]
@slawekjaranowski I have pushed a naive solution to the problem. @elharo WDYT?

File: src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java (hunk 1 above)
Comment (elharo) [reply src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java]: See section 2.2 and 4.1 of the XML 1,.0 spec: https://www.w3.org/TR/REC-xml/#charsets
"Characters referred to using character references MUST match the production for [Char](https://www.w3.org/TR/REC-xml/#NT-Char)."

XML 1,1 is a little more permissive, but still prohibits the NUL character.

Comment (michael-o) [reply src/test/java/org/codehaus/plexus/util/xml/Xpp3DomWriterTest.java]: Yes, noticed as well. My approach is not correct.

**Comment by slawekjaranowski** [comment]
Next commit with removing special chars - fix build in invoker

**Review by michael-o — APPROVED** [review]
Looks reasonable, but I'd like @elharo to check as well.

**Comment by michael-o** [comment]
Can 4.x be used with Maven 3.x?

**Comment by slawekjaranowski** [comment]
> Can 4.x be used with Maven 3.x?

Generally no - I will cherry pick to 3.x

File: src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    private String expectedOutput() {
+        StringBuilder out = new StringBuilder();
+        out.append("<?xml version=\"1.0\" standalone=\"yes\"?>");
+        out.append("<root>");
+        out.append("<char>BACKSPACE: </char>");
+        out.append("<char>CHARACTER TABULATION: \t");
+        out.append("</char><char>LINE FEED (LF): \n");
+        out.append("</char><char>LINE TABULATION: </char>");
+        out.append("<char>CARRIAGE RETURN (CR): \r");
+        out.append("</char><char>SHIFT IN: </char>");
```
Comment (elharo) [inline src/test/java/org/codehaus/plexus/util/xml/pull/MXSerializerTest.java:54]: suggest moving first `</char>` to previous string

File: src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java [hunk 3]
```diff
@@ -228,8 +228,10 @@ private static String escapeXml(String text) {
 
     private static final Pattern lowers = Pattern.compile("([\000-\037])");
 
+    private static final Pattern lowersText = Pattern.compile("([\000-\010\013-\014\016-\037])");
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java:231]: Personally I find hexadecimal codes easier to follow than octal. I had to look these up. 

Also "lowersText" isn't obvious to me. Perhaps "illegalC0Characters"

File: src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java [hunk 4]
```diff
@@ -247,6 +249,19 @@ private static String escapeXmlAttribute(String text) {
         return b.toString();
     }
 
+    private static String escapeXmlText(String text) {
+        text = escapeXml(text);
+
+        Matcher m = lowersText.matcher(text);
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java:255]: m -=-> matcher

File: src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java [hunk 5]
```diff
@@ -247,6 +249,19 @@ private static String escapeXmlAttribute(String text) {
...
+        StringBuffer b = new StringBuffer();
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java:256]: StringBuilder builder

**Review by elharo — COMMENTED** [review]
I'm not very familiar with this API or code base, but at first glance it seems OK. I would add documentation in the appropriate spots indicating that C0 controls are omitted from output.

File: src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java (hunk 5 above)
Comment (slawekjaranowski) [reply src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java:256]: `appendReplacement` require `StringBuffer`, `StringBuilder` is supported from JDK 9

**Review by michael-o — APPROVED** [review]

File: src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java [hunk 6]
```diff
@@ -24,9 +24,9 @@
 import java.util.regex.Pattern;
 
 /**
- * Implementation of XMLWriter which emits nicely formatted documents.
- *
+ * <p>Implementation of XMLWriter which emits nicely formatted documents.</p>
  *
+ * <p>C0 controls chars are omitted from output</p>
```
Comment (elharo) [inline src/main/java/org/codehaus/plexus/util/xml/PrettyPrintXMLWriter.java:29]: C0n control characters except \n, \r, and \t

**Review by elharo — COMMENTED** [review]
It occurs to me that there are also some issues in this class with unpaired surrogates, but one piece of technical arcana at a time. :-)

**Review by elharo — APPROVED** [review]

**Pull request merged**